### PR TITLE
CON-638 Experimental developer support for external ES5

### PR DESCRIPTION
**Pull request by lancewf**
This adds elastic search 5 support to chef_index but this support is
experimental.  It does not take care of upgrading a current chef-server
from ES 2 -> ES 5.  It also does not ensure that all other chef-server
functionality such as the reporting or the chef-backend plugin correctly
work when ES 5 is used.  Using ES5 as the search engine is completely
unsupported in a customer environment.

For the dev environment, a new optional VM has been added to
dev/Vagrantfile.  To enable it, check out defaults.yml for the commented
out 'elasticsearch' section.  It supports bring ES2 or ES5 currently.

@ssd, @sdelano, @marcparadise, @chef/chef-server-maintainers

### CODE REVIEW

File: omnibus/files/private-chef-cookbooks/private-chef/libraries/helper.rb [hunk 1]
```diff
@@ -58,6 +59,28 @@ def solr_root
...
+      elastic_search_url = node['private_chef']['opscode-solr4']['external_url']
+      begin
+        elastic_search_uri = URI.parse(elastic_search_url)
+        req = Net::HTTP::Get.new(elastic_search_uri)
+        res = Net::HTTP.start(elastic_search_uri.hostname, elastic_search_uri.port, use_ssl: elastic_search_uri.scheme == "https") do |http|
+          http.request(req)
+        end
+        version = JSON.parse(res.body)['version']['number'].split('.').first.to_i
+        version = default_version if version == 0
+        version
```
Comment (stevendanna) [inline omnibus/files/private-chef-cookbooks/private-chef/libraries/helper.rb]: Should we check that the discovered version is one we support?

File: dev/scripts/provision-elasticsearch.sh [hunk 2]
```diff
@@ -0,0 +1,17 @@
+add-apt-repository ppa:openjdk-r/ppa
```
Comment (ksubrama) [inline dev/scripts/provision-elasticsearch.sh]: Yeah makes sense.  I was just doing copy-pasta based on the surrounding code for postgres and chef-backend to make my life easier in the Vagrantfile.

File: omnibus/files/private-chef-cookbooks/private-chef/libraries/helper.rb (hunk 1 above)
Comment (ksubrama) [inline omnibus/files/private-chef-cookbooks/private-chef/libraries/helper.rb]: I considered that - but then I didn't know what we did support.  There's only ES 1, 2 and 5 right now (6 is on the way).  I guess we don't support 1.  Are there any minor versions in 2 that we don't support?

File: omnibus/files/private-chef-cookbooks/private-chef/recipes/opscode-erchef.rb [hunk 3]
```diff
@@ -33,6 +33,8 @@
 
 rabbitmq = OmnibusHelper.new(node).rabbitmq_configuration
 
+elastic_search_version = OmnibusHelper.new(node).elastic_search_major_version
```
Comment (ksubrama) [inline omnibus/files/private-chef-cookbooks/private-chef/recipes/opscode-erchef.rb]: It's being passed below in line 57.

File: omnibus/files/private-chef-cookbooks/private-chef/templates/default/pedant_config.rb.erb [hunk 4]
```diff
@@ -75,7 +75,7 @@ search_server "<%= @solr_url %>"
 
 <% if node['private_chef']['opscode-erchef']['search_provider'] == "elasticsearch" %>
 search_commit_url "/_refresh"
-search_url_fmt "/chef/_search?fq=X_CHEF_type_CHEF_X:%{type}&q=%{query}"
+search_url_fmt "/chef/_search?q=X_CHEF_type_CHEF_X:%{type}%%20%{query}"
```
Comment (ksubrama) [inline omnibus/files/private-chef-cookbooks/private-chef/templates/default/pedant_config.rb.erb]: Elastic search experts: Can you confirm if this is a legit change.  This is what @lancewf and I came up with to get the current tests passing but maybe there's an edge case we're missing.

File: src/oc_erchef/apps/chef_index/README.md [hunk 5]
```diff
@@ -6,7 +6,7 @@ Chef is a system integration framework written in erlang and ruby and designed t
 
 The Chef Wiki is the definitive source of user documentation.
 
-    http://wiki.chef.io/display/chef/Home
```
Comment (ksubrama) [inline src/oc_erchef/apps/chef_index/README.md]: The only line I think is important is the "put any changes that affect oc_reporting in chef_index".  I'll trim the rest.

Speaking of which...  I probably need to port some of this over don't I?

Comment (stevendanna) [inline src/oc_erchef/apps/chef_index/README.md]: If and only if we care about reporting support on ES5 :D

File: omnibus/files/private-chef-cookbooks/private-chef/recipes/opscode-erchef.rb (hunk 3 above)
Comment (stevendanna) [inline omnibus/files/private-chef-cookbooks/private-chef/recipes/opscode-erchef.rb]: In my diff, line 57 is literally a new empty line, is github trolling me?

Comment (ksubrama) [inline omnibus/files/private-chef-cookbooks/private-chef/recipes/opscode-erchef.rb]: Figured it out.  @lancewf 's editor had been trolling him the entire time we coded this up.  It put in 10 tabs or something instead of 20 spaces and I'm fairly sure github uses 4 or 8 space tabs.

File: src/oc_erchef/apps/chef_index/README.md (hunk 5 above)
Comment (ksubrama) [inline src/oc_erchef/apps/chef_index/README.md]: Oh good point.  Yeah f-that.  We should have automate notifications in a good enough place to ask anyone upgrading to ES5 to use that instead of reporting. #famouslastwords.

Heads up to @charlesjohnson.

File: omnibus/files/private-chef-cookbooks/private-chef/recipes/opscode-erchef.rb (hunk 3 above)
Comment (lancewf) [inline omnibus/files/private-chef-cookbooks/private-chef/recipes/opscode-erchef.rb]: I am not sure how that happened. We even tested that change before we checked it in.

File: omnibus/files/private-chef-cookbooks/private-chef/libraries/helper.rb (hunk 1 above)
Comment (lancewf) [inline omnibus/files/private-chef-cookbooks/private-chef/libraries/helper.rb]: I updated this part to log an error if the major version is not 2 or 5.

File: omnibus/files/private-chef-cookbooks/private-chef/recipes/opscode-erchef.rb (hunk 3 above)
Comment (marcparadise) [inline omnibus/files/private-chef-cookbooks/private-chef/recipes/opscode-erchef.rb]: "Worked on my machine - it's an ~ops~ pipeline problem now!" :D

**Comment by marcparadise** [comment]
A couple of things we'll need to verify for this: 

* Reporting uses a forked chef_index search interface. It does not currently support ES, but a sanity check to ensure that reporting pedant tests pass would be good - the path that uses search will be hit in those tests.  
* We're building this because we need it, but what do we want to do in terms of documenting the change for customers? This is a factor for cloud all-in-one installations , but also customers who are on the "backendless" path.

**Comment by ksubrama** [comment]
We added some unit tests to the library in the omnibus cookbook.  I figure that the pedant tests exercising functionality should cover most of the other scenarios we care about and ensure the ES2/5 work like we expect solr to work.  It isn't automated at all - but I have run pedant tests locally for both ES2 and ES5.  There's one test that involves rabbitmq that seems to be flaking out but that test behaves the same way in master as well - so I'm not worrying about it.  All the other tests past.

I'll run the reporting pedant tests and see if that still works on just solr.

**Comment by kmacgugan** [comment]
The query changes look good to me.

**Comment by ksubrama** [comment]
Ran opscode-reporting-ctl test --all using solr on this branch!  Everything still passes.

File: omnibus/files/private-chef-cookbooks/private-chef/libraries/helper.rb [hunk 6]
```diff
@@ -58,6 +59,31 @@ def solr_root
     host
   end
 
+  def elastic_search_major_version
```
Comment (stevendanna) [inline omnibus/files/private-chef-cookbooks/private-chef/libraries/helper.rb]: I'm a bit concerned that a transient ES issue during `reconfigure` could rather silently (one log message in hundreds of lines of output) reconfigure us to use version 2 instead of version 5 and produce very confusing errors for the user.

Comment (ksubrama) [inline omnibus/files/private-chef-cookbooks/private-chef/libraries/helper.rb]: That's a good point...  Should we raise exceptions and fail hard for some of these?

Comment (ksubrama) [inline omnibus/files/private-chef-cookbooks/private-chef/libraries/helper.rb]: Fixed to raise exceptions when faced with invalid data.  Also does retries on flaky connections.
